driver_integration_test.go

```diff
@@ -301,3 +301,22 @@ func TestServiceAccountAuthentication(t *testing.T) {
 		t.Errorf("Authentication didn't return an error with correct message, got: %s", err.Error())
 	}
 }
+
+func TestIncorrectQueryThrowingStructuredError(t *testing.T) {
+	db, err := sql.Open("firebolt", dsnSystemEngineMock)
+	if err != nil {
+		t.Errorf("failed unexpectedly with %v", err)
+	}
+	_, err = db.Query("SET advanced_mode=1")
+	raiseIfError(t, err)
+	_, err = db.Query("SET enable_json_error_output_format=true")
+	raiseIfError(t, err)
+	_, err = db.Query("SELECT 'blue'::int")
+	if err == nil {
```

Comment:
Also verify that err is a StructuredError, otherwise it seems like plain error here whould also pass

Reply:
Unfortunately in go our error class becomes masked due to the `ConstructNestedError` which converts the error to string before throwing it up the stack.
However, looking through the code other error paths prepend their own prefixes. e.g. `request returned non ok status code` if the error code is not 200 and `request returned an error` in some other cases. Essentially in this test by checking the exact prefix is `error during query execution: error during query request: Cannot parse string` we verify the error path here by the process of elimination.
I don't think we can do better here, unless we prepend a specific prefix similar to the above. Not sure if we should do it as our errors are already pretty verbose.

Reply:
As we discussed, I believe we can be sure that if engine returns a json string and we don't parse it, the test will fail. This test is only false positive in case the backend still returns a plaintext despite of the setting, which I believe we might not account for in our testing. We can probably leave it as is

Reply:
> This test is only false positive in case the backend still returns a plaintext despite of the setting

I don't think it will, since the StructuredError will not be raised so the error will be raised elsewhere in the code e.g. [here](https://github.com/firebolt-db/firebolt-go-sdk/blob/95503e01398bd3c36d10ab3474784ab92bb5db3c/client_base.go#L294) we prepend the `request returned non ok status code` and in the [test](https://github.com/firebolt-db/firebolt-go-sdk/pull/108/files/95503e01398bd3c36d10ab3474784ab92bb5db3c#diff-d9302a33e0d5b3d05467bda38b444952255f4e799fca7079fd9bc6c8255e60bfR319) we specifically check for a prefix that doesn't contain this string. Therefore I believe the plain error will make this test fail.
In fact, I've removed the set parameters, run it and got an error:
```
--- FAIL: TestIncorrectQueryThrowingStructuredError (0.11s)
    /Users/petrotiurin/Development/firebolt-go-sdk/driver_integration_test.go:320: Query didn't return an error with correct message, got: error during query execution: error during query request: request returned non ok status code: 400, Cannot parse string 'blue' as integer: syntax error at beginning of string
FAIL
```